inmoose/edgepy/edgepy_cpp/edgepy_cpp.pyx

```diff
@@ -77,3 +88,80 @@ cdef public void f77_dpotrs "f77_dpotrs"(const char *uplo, const int *n, const i
     dpotrs(<char*>uplo, <int*>n, <int*>nrhs, <double*>a, <int*>lda, b, <int*>ldb, info)
 cdef public void f77_dsytrf "f77_dsytrf"(const char *uplo, const int *n, double *a, const int *lda, int *ipiv, double *work, const int *lwork, int *info) nogil:
     dsytrf(<char*>uplo, <int*>n, a, <int*>lda, ipiv, work, <int*>lwork, info)
+
+
+cdef double mildly_low_value = 1e-8
+@cython.ufunc
+@cython.cdivision(True)
+cdef double compute_unit_nb_deviance(double y, double mu, double phi):
+    """
+    Calculate the deviance of a negative binomial fit
+
+    Note the protection for very large mu*phi (where we use a Gamma instead) or
+    very small mu*phi (where we use a Poisson instead). This approximation
+    protects against numerical instability introduced by subtracting a very
+    large log value in (log mu) with another very large logarithm (log mu+1/phi).
+    We need to consider the phi as the approximation is only good when the
+    product is very big or very small.
+
+    Arguments
+    ---------
+    y : array_like
+        counts matrix
+    mu : array_like
+        expected means matrix (broadcastable to the shape of :code:`y`)
+    phi : array_like
+        dispersion matrix (broadcastable to the shape of :code:`y`)
+
+    Returns
+    -------
+    ndarray
+        matrix of deviances (same shape as :code:`y`)
+    """
+    # add a small value to protect against zero during division and log
+    y = y + mildly_low_value
+    mu = mu + mildly_low_value
+
+    # Calculating the deviance using either the Poisson (small phi*mu), the
+    # Gamma (large phi*mu) or NB (everything else). Some additional work is
+    # put in to make the transitiosn between families smooth
+    resid = y - mu
+    product = mu * phi
+
+    if phi < 1e-4:
```

Comment:
How can an array be smaller than a single value ?

Reply:
This function is a UFunc (see the `cython.ufunc` decorator): its code describe what to do on individual values, and Cython takes care of the vectorization to call it on arrays.
So `phi` here is a `double` (as per the function signature), although the docstring says it is an array.